src/hypothesis/searchstrategy/datetime.py

```diff
@@ -0,0 +1,98 @@
+# coding=utf-8
+#
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis-python
+#
+# Most of this work is copyright (C) 2013-2017 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at http://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+from __future__ import division, print_function, absolute_import
+
+import datetime as dt
+
+from hypothesis.errors import NoExamples
+from hypothesis.internal.conjecture import utils
+from hypothesis.searchstrategy.strategies import SearchStrategy
+
+__all__ = ['DatetimeStrategy', 'TimedeltaStrategy']
+
+
+class DatetimeStrategy(SearchStrategy):
+
+    def __init__(self, min_datetime, max_datetime, timezones_strat):
+        assert isinstance(min_datetime, dt.datetime)
+        assert isinstance(max_datetime, dt.datetime)
+        assert min_datetime.tzinfo is None
+        assert max_datetime.tzinfo is None
+        assert min_datetime <= max_datetime
+        assert isinstance(timezones_strat, SearchStrategy)
+        self.min_dt = min_datetime
+        self.max_dt = max_datetime
+        self.tz_strat = timezones_strat
+
+    def do_draw(self, data):
+        # I've never seen this take eight iterations, but let's be safe...
```

Comment:
I think this can break in weird ways when things are being mutated and/or shrunk. It's important to remember that a ConjectureData object is not actually a random number generator and can, by design, fuck around arbitrarily with the data it generates.

The comparable logic in filter is [to try 3 times and then fail](https://github.com/HypothesisWorks/hypothesis-python/blob/master/src/hypothesis/searchstrategy/strategies.py#L313) which might be better to emulate here (though three is an arbitrary number and it could be 8 just as easily - filter might be expensive so it makes sense to keep it smaller).

Failing (by calling `mark_invalid()` on the data object) is definitely the correct thing to do at the end of the loop rather than raising an exception.

Regardless of the number you end up settling on, here's how you might test this:

1. Factor out the loop body into its own method.
2. Use `strategies.binary()` + `find()` to find a buffer such that when `do_draw` is called on `ConjectureData.for_buffer` the loop body fails. 
3. Call `data.draw(datetimes())` on a `ConjectureData.for_buffer(failure_inducing * 100)`.